network/simulation/bucket_test.go

```diff
@@ -82,22 +76,18 @@ func TestServiceBucket(t *testing.T) {
 	t.Run("SetNodeItem", func(t *testing.T) {
 		sim.SetNodeItem(id1, customKey, customValue)
 
-		v, ok := sim.NodeItem(id1, customKey)
-		if !ok {
-			t.Fatal("bucket item not found")
-		}
-		s, ok := v.(string)
-		if !ok {
-			t.Fatal("bucket item value is not string")
-		}
+		v := sim.NodeItem(id1, customKey)
+		s := v.(string)
 		if s != customValue {
 			t.Fatalf("expected %q, got %q", customValue, s)
 		}
 
-		_, ok = sim.NodeItem(id2, customKey)
-		if ok {
-			t.Fatal("bucket item should not be found")
-		}
+		defer func() {
+			if r := recover(); r == nil {
```

Comment:
This is rather weird - you seem to modify `NodeItem` to panic and not return errors, and here you are catching all the panic.

1. I suggest we either keep the interface as is, and check for errors, or,
2. If this is used only in test, pass `t` as an argument, add a prefix `Must` to the method name, and `t.Fatal` inside the method. Then it would be `t.Fatal`, and not `panic`, and you wouldn't have to check for errors every time you use is.

Reply:
`NodeItem` is only used in tests, so if we assume we don't need to use it outside of the `go test` suite, you could probably add a `MustNodeItem (t *testing.T, ...)` and `t.Fatal` inside in case of errors.

Reply:
i did this because of this annoying boilerplate of error checks everywhere, although they definitely shouldn't happen (they happen only when the entry does not exist in the bucket. i've never seen this happen).
what's the problem with having a panic here?
the test that i changed above is recovering from the panic because this is exactly what the original test was checking for - returning an error for an entry that doesn't exist. instead - i'm panicking and here catching the panic in order to verify it was indeed thrown...

Reply:
`panic` means that something should never happen and you as developer don't expect the code to be used in a way that will result in the panic.

if you are catching the panic, it means that you expect it to happen, and are hiding the interface - if another developer uses the same code, they wouldn't know they have to use `recover` to account for these failures.

this is why we return `errors` as part of the interface.

Reply:
i'm not sure i'm following... the defer statement catches just the one panic that is supposed to happen, because the previous test was checking that the item should not be found, hence returning an error from the `NodeItem` call. now because i prefer this to panic - I'm catching the panic in order to assert that it indeed happened. if the test panics earlier in the code - this will not be caught by the recover statement (since the defer statement is called only before the statement that is supposed to fail).
I really prefer to not have the returned error here. It practically should not happen at all, that's why I want it to panic

Reply:
> if you are catching the panic, it means that you expect it to happen, and are hiding the interface - if another developer uses the same code, they wouldn't know they have to use recover to account for these failures.

the panic should happen when the item is not found and that should immediately indicate to the developer that he/she fucked up something. no one is expected to `recover` from anything, just as you are not supposed to recover from doing a cast on a type that panics

Reply:
You've basically replaces returned errors with panics, and assume that everyone will be using the `NodeItem` as you - expecting all the items they check to be there. There can be many tests that just confirm that an item is NOT there, and they'd have to then be catching the panics.

So I suggest we revert back to this interface returning errors so that it can be used when items are not there, and if you want to keep your code clean and not check 100 times for an error, just add another one called `MustNodeItem` which takes `t *testing.T` as argument and fails if an item is not there.

No need for panics in this case at all.

Reply:
> There can be many tests that just confirm that an item is NOT there

There are _ZERO_ tests that check that the item is not there.
If you still insist I'll revert this but please make sure you look at current master and at ALL cases on how this function is being used. thank you